examples/cms/src/FieldMapping.tsx

```diff
@@ -0,0 +1,223 @@
+    return (
+        <>
+            <div
+                className="source-field"
+                aria-disabled={isIgnored}
+                onClick={() => onFieldToggle(field.id)}
+                role="button"
+            >
+                <input
+                    type="checkbox"
+                    checked={!isIgnored}
+                    onChange={event => {
+                        event.stopPropagation()
+                    }}
+                />
+                <span>{field.id}</span>
+            </div>
+            <ChevronIcon />
+            <input
+                type="text"
+                style={{
+                    width: "100%",
+                    opacity: isIgnored ? 0.5 : 1,
+                }}
+                disabled={isIgnored}
+                placeholder={field.id}
+                value={field.name}
+                onChange={event => onFieldNameChange(field.id, event.target.value)}
+                onKeyDown={event => {
+                    if (event.key === "Enter") {
+                        event.preventDefault()
+                    }
+                }}
+            />
+        </>
+    )
+})
+
+const managedCollection = await framer.getManagedCollection()
+const initialManagedCollectionFields: ManagedCollectionFields = await managedCollection.getFields()
+const initialFieldIds: FieldIds = new Set()
+
+interface FieldMappingProps {
+    collection: ManagedCollection
+    dataSource: DataSource
+    initialSlugFieldId: string | null
+}
+
+export function FieldMapping({ collection, dataSource, initialSlugFieldId }: FieldMappingProps) {
+    const [status, setStatus] = useState<"mapping-fields" | "loading-fields" | "syncing-collection">("loading-fields")
+    const isSyncing = status === "syncing-collection"
+    const isLoadingFields = status === "loading-fields"
+
+    const sourceFields = useMemo(() => computeFieldsFromDataSource(dataSource), [dataSource])
+    const possibleSlugFields = useMemo(() => sourceFields.filter(field => field.type === "string"), [sourceFields])
+
+    const [selectedSlugField, setSelectedSlugField] = useState<ManagedCollectionField | null>(
+        possibleSlugFields.find(field => field.id === initialSlugFieldId) ?? possibleSlugFields[0] ?? null
+    )
+
+    const [fields, setFields] = useState(initialManagedCollectionFields)
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState(initialFieldIds)
+
+    useEffect(() => {
+        setFields(mergeFieldsWithExistingFields(sourceFields, initialManagedCollectionFields))
+
+        const existingFieldIds = new Set(initialManagedCollectionFields.map(field => field.id))
+        const ignoredFields = sourceFields.filter(sourceField => !existingFieldIds.has(sourceField.id))
+
+        if (initialSlugFieldId) {
+            setIgnoredFieldIds(new Set(ignoredFields.map(field => field.id)))
+        }
+
+        setStatus("mapping-fields")
```

Comment:
I think this used to be an effect because it depended on data that wasn't available yet. But currently this could be calculated outside of an effect I think. Probably because you are loading the entire collection and field outside the component?